mixing/mixpool/mixpool.go

```diff
@@ -0,0 +1,1523 @@
+			for hash := range r.hashes {
+				delete(p.pool, hash)
+			}
+		}
+	}
+
+	// Expire PRs and remove identity tracking
+	for _, pr := range p.prs {
+		if pr.Expiry > height {
+			continue
+		}
+
+		p.removePR(pr, "expired")
+	}
+
+	// Expire orphans with old receive times, and in the case of any
+	// orphan KE, expire those with old epochs.
+	for hash, o := range p.orphans {
+		expire := time.Since(o.accepted) >= 20*time.Minute
+		if !expire {
+			if ke, ok := o.message.(*wire.MsgMixKeyExchange); ok {
+				epoch := time.Unix(int64(ke.Epoch), 0)
+				expire = time.Since(epoch) >= 20*time.Minute
+			}
+		}
+		if expire {
+			delete(p.orphans, hash)
+			delete(p.orphansByID, *(*[33]byte)(o.message.Pub()))
+		}
+	}
+}
+
+// RemoveMessage removes a message that was rejected by the network.
+func (p *Pool) RemoveMessage(msg mixing.Message) {
+	p.mtx.Lock()
+	defer p.mtx.Unlock()
+
+	msgHash := msg.Hash()
+	delete(p.pool, msgHash)
+	if pr, ok := msg.(*wire.MsgMixPairReq); ok {
+		p.removePR(pr, "rejected")
+	}
+	if ke, ok := msg.(*wire.MsgMixKeyExchange); ok {
+		delete(p.latestKE, ke.Identity)
+	}
+}
+
+// RemoveSession removes all non-PR messages from a completed or errored
+// session.  PR messages of a successful run (or rerun) are also removed.
+func (p *Pool) RemoveSession(sid [32]byte, success bool) {
+	p.mtx.Lock()
+	defer p.mtx.Unlock()
+
+	p.removeSession(sid, nil, success)
+}
+
+func (p *Pool) removeSession(sid [32]byte, txHash *chainhash.Hash, success bool) {
+	ses := p.sessions[sid]
+	if ses == nil {
+		return
+	}
+
+	// Delete PRs used to form final run
+	var removePRs []chainhash.Hash
+	var lastRun *runstate
+	if success {
+		lastRun = &ses.runs[len(ses.runs)-1]
+		removePRs = lastRun.prs
+	}
+
+	if txHash != nil || success {
+		if txHash == nil {
+			// XXX: may be better to store this in the runstate as
+			// a CM is received.
+			for h := range lastRun.hashes {
+				if e, ok := p.pool[h]; ok && e.msgtype == msgtypeCM {
+					cm := e.msg.(*wire.MsgMixConfirm)
+					hash := cm.Mix.TxHash()
+					txHash = &hash
+					break
+				}
+			}
+		}
+		if txHash != nil {
+			delete(p.sessionsByTxHash, *txHash)
+		}
+	}
+
+	delete(p.sessions, sid)
+	for _, r := range ses.runs {
+		for hash := range r.hashes {
+			delete(p.pool, hash)
+		}
+	}
+
+	for _, prHash := range removePRs {
+		delete(p.pool, prHash)
+		if pr := p.prs[prHash]; pr != nil {
+			p.removePR(pr, "mixed")
+		}
+	}
+}
+
+// RemoveConfirmedRuns removes all messages including pair requests from
+// runs which ended in each peer sending a confirm mix message.
+func (p *Pool) RemoveConfirmedRuns() {
+	p.mtx.Lock()
+	defer p.mtx.Unlock()
+
+	p.removeConfirmedRuns()
+}
+
+func (p *Pool) removeConfirmedRuns() {
+	for sid, ses := range p.sessions {
+		lastRun := &ses.runs[len(ses.runs)-1]
+		cmCount := lastRun.countFor(msgtypeCM)
+		if len(lastRun.prs) != int(cmCount) {
+			continue
+		}
+
+		delete(p.sessions, sid)
+		for _, prevRun := range ses.runs[:len(ses.runs)-1] {
```

Comment:
Any reason why this does the same thing separately for previous runs and the last run?

It seems like it could just be:
```Go
for _, run := range ses.runs {
	for hash := range prevRun.hashes {
	delete(p.pool, hash)
	}
}
```

Reply:
does this now